python/lib/communication/dmod/communication/maas_request/ngen/partial_realization_config.py

```diff
@@ -0,0 +1,68 @@
+from ngen.config.configurations import Routing
+from ngen.config.formulation import Formulation
+from ngen.config.realization import CatchmentRealization
+from pydantic import BaseModel, validator
+from typing import ClassVar, Dict, List, Optional
+
+
+class PartialRealizationConfig(BaseModel):
+    """
+    Helper class for working with the serialized formulation configurations ::class:`AbstractNgenRequest` messages.
+
+    The type relies on classes from the ``ngen-config`` external package and the transitive ``pydantic`` dependency.
+
+    Note that this type has two optional forcing-config-related fields. These are expected to be used infrequently, as
+    generally DMOD will automatically determine the correct forcing dataset. They do allow for optional user control,
+    but the more specific and essential need (which does rely on such control being available) is to support a
+    workaround for reading forcing data directly from the Docker host via a bind mount. This functionality is handled by
+    the scheduling logic, but the generated realization config must properly reflect it.
+
+    The workaround is triggered by adding specific prefix (::attribute:`_FROM_ENV_PREFIX`) and delimiter
+    (::attribute:`_FROM_ENV_DELIMIT`) substrings to either of the forcing config fields.
+    """
+
+    _FROM_ENV_DELIMIT: ClassVar[str] = ':::'
+    _FROM_ENV_PREFIX: ClassVar[str] = 'from_env'
+
+    hydrofabric_uid: str
+    """ The unique id of hydrofabric associated with the catchments to which the contained formulations apply. """
+
+    global_formulations: Optional[List[Formulation]] = None
+    """ The global formulation(s) config, serving as a default once in a full NextGen realization configuration. """
+
+    catchment_formulations: Optional[Dict[str, CatchmentRealization]] = None
+    """ The individual catchment formulation configs, if set, keyed by catchment id. """
+
+    forcing_file_pattern: Optional[str] = None
+    """ Optional catchment-id-based pattern string for basename of per-catchment forcing files. """
+
+    forcing_file_name: Optional[str] = None
+    """ Optional fixed name for the forcing data file. """
+
+    routing_config: Optional[Routing] = None
+    """ Optional routing config object for the partial config. """
+
+    is_env_workaround: bool = None
```

Comment:
`None` or `False`?

Reply:
I think we actually do want `None` here as a hint that other things are in place to make sure the value is always a _bool_.  The `default_is_env_workaround` validator function should make sure that _something_ other than `None` is always provided for `is_env_workaround`.

That's a style thing, though, so I'd be open to changing it in the future if we need.

Reply:
> I think we actually do want `None` here as a hint that other things are in place to make sure the value is always a _bool_.

We should check that. I think because you used `always=True` and `pre=True` this can result in `None`.